primitives/authority-selection-inherents/src/authority_selection_inputs.rs

```diff
@@ -11,14 +11,33 @@ use {
 	main_chain_follower_api::DataSourceError, sidechain_domain::McEpochNumber,
 };
 
-/// The part of data for selection of authorities that comes from the main chain.
+/// The part of data for selection of authorities that comes from the main chains.
 /// It is unfiltered, so the selection algorithm should filter out invalid candidates.
+/// Note, this is part of the state storage (see pallet_session_validator_management).
 #[derive(Clone, Debug, Encode, Decode, TypeInfo, PartialEq, Eq)]
 pub struct AuthoritySelectionInputs {
 	pub d_parameter: DParameter,
+
+	/// List of permissioned candidates from the side chain, one for each sidechain_public_key.
 	pub permissioned_candidates: Vec<PermissionedCandidateData>,
+
+	// TODO ETH: The items in the Vec should be uniquely identified by sidechain_public_key
+	// (instead of  mainchain_public_key). This is necessary to allow one PC operator to register
+	// stake on different chains. The registration schema is symmetric with all chains and hence
+	// the same sidechain_public_key can be related to many public keys on different chains.
+	// Thus:
+	// - only one CandidateRegistrations per each candidate
+	// - each registered_candidates[i]:
+	//   - should contain valid mainchain_pub_key
+    //   - may have optional eth_pub_key, in which case it may also have valid EthRegistrationData records
+
+	/// List of registered candidates from the main chain, one for each sidechain_public_key.
 	pub registered_candidates: Vec<CandidateRegistrations>,
+
+	/// The nonce for the epoch coming from the main chain.
 	pub epoch_nonce: EpochNonce,
+
+	// TODO ETH: add nonce from Ethereum chain
```

Comment:
I think that either we target 3 sources and hardcode it in this way, so we have permissioned_candidates, registered_cardano_candidates, registered_ethereum_candidates or this code should be generic for any number of sources, where each source has its "score" and each candidate inside this candidates source has it's score relative to other candidates of this type.

Said that I don't see Minotaur algorithm and I can only guess who it is supposed to be generalization of Ariadne.

Reply:
I'm strongly for generalization here. I don't think that this part of the code has to know anything about the name of the source of candidates.

Reply:
@aslesarenko lets chat more about this one.  given we now have 2 different tokens, makes sense to make generic here.